File: src/types/codexJson.ts
/** codex exec --json 杈撳嚭鐨?JSONL 浜嬩欢 */

export type CodexJsonEvent =
  | { type: 'thread.started'; thread_id: string }
  | { type: 'turn.started' }
  | { type: 'item.started'; item: CodexItem }
  | { type: 'item.completed'; item: CodexItem }
  | { type: 'turn.completed'; usage?: CodexUsage }
  | { type: 'item.agent_message.delta'; item_id: string | null; delta: string }
  | { type: 'item.reasoning.delta'; item_id: string | null; delta: string }
  | { type: 'item.reasoning.summary.delta'; item_id: string | null; delta: string }
  | { type: 'turn.plan.updated'; explanation?: string | null; plan: TurnPlanStep[] }
  | { type: 'turn.diff.updated'; diff: string }
  | { type: 'error'; message: string }

/** codex 杈撳嚭鐨勯」鐩?*/
export type CodexItem =
  | { id: string; type: 'error'; message: string }
  | { id: string; type: 'agent_message'; text: string }
  | {
      id: string
      type: 'tool_call'
      name: string
      arguments?: unknown
      status?: string
    }
  | {
      id: string
      type: 'mcp_tool_call'
      server?: string
      tool?: string
      arguments?: unknown
      result?: unknown
      status?: 'in_progress' | 'completed' | 'failed'
      error?: { message?: string } | null
    }
  | {
      id: string
      type: 'approval_request'
      command?: string
      description?: string
      /** 瑙勫垯寮曟搸鍐崇瓥锛歛sk / auto_accept / auto_deny */
      decision?: string
      reason?: string
    }
  | {
      id?: string
      type: 'command_execution'
      command?: string
      status?: 'in_progress' | 'completed'
      aggregated_output?: string
      exit_code?: number | null
    }
  | {
      id: string
      type: 'file_change'
      changes: CodexFileChange[]
      status?: string
    }

/** token 浣跨敤缁熻?*/
export interface CodexUsage {
  input_tokens?: number
  cached_input_tokens?: number
  cache_write_input_tokens?: number
  output_tokens?: number
  reasoning_output_tokens?: number
}

/** 鏂囦欢鍙樻洿椤癸紙file_change item 鍐呯殑鍗曚釜鏂囦欢锛?*/
export interface CodexFileChange {
  path: string
  kind: 'add' | 'delete' | 'update'
}

/** 鍓嶇湪娓叉煋鐢ㄧ殑缁撴瀯鍖栨秷鎭?*/
export interface CodexMessage {
  id: string
  kind:
    | 'user'
    | 'agent'
    | 'tool'
    | 'error'
    | 'system'
    | 'reasoning'
    | 'subagent'
    | 'usage'
    | 'file_change'
    | 'plan'
    | 'review'
    | 'deny'
    | 'turn_summary'
    | 'reflection'
  content: string
  toolName?: string
  toolArgs?: unknown
  usage?: CodexUsage
  /** file_change 娑堟伅鐨勬枃浠跺彉鏇村垪琛?*/
  fileChanges?: CodexFileChange[]
  /** deny 鍗″榻灞曠ず鐨勫懡涓浠ュ師鍥狅紙鐢ㄦ埛瑙勫垯/鍐呯疆瑙勫垯璇存槑锛?*/
  reason?: string
  /** 宸ュ叿/MCP 璋冪敤鑰楁椂锛堟椃绉掞級 */
  durationMs?: number
  /** 宸ュ叿/MCP 璋冪敤鐨勭粨鏋滄憳瑕侊紙棣栬党建200瀛楋級 */
  toolResult?: string
  /** 鏈瑁″熀鏁帮紝浠呮Turn_summary 浣跨敤锛?*/
  turnStats?: TurnStats
  timestamp: number
}

/** 鏈瑁″仠浣缁熻锛坱urn_summary 娑堟伅浣跨敤锛?*/
export interface TurnStats {
  durationMs: number
  toolCalls: number
  mcpCalls: number
  fileChanges: number
  hadErrors: boolean
  inputTokens: number
  outputTokens: number
  reasoningTokens: number
  cacheReadTokens: number
  cacheWriteTokens: number
}

/** 模型任务清单的一步(来自 codex `turn/plan/updated`,对齐 Claude Code 的 TodoWrite) */
export interface TurnPlanStep {
  step: string
  status: 'pending' | 'inProgress' | 'completed'
}

File: src-tauri/src/types/codex_json.rs
use serde::{Deserialize, Serialize};

/// codex exec --json 输出的 JSONL 事件
///
/// 每一行是一个 JSON 对象，type 字段标识事件类型。
/// 注意：type 值带点号（如 "thread.started"），不能用 rename_all，必须显式 rename。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum CodexJsonEvent {
    /// 会话开始
    #[serde(rename = "thread.started")]
    ThreadStarted {
        #[serde(rename = "thread_id")]
        thread_id: String,
    },
    /// 一轮对话开始
    #[serde(rename = "turn.started")]
    TurnStarted,
    /// 某个项目完成（消息、工具调用、错误等）
    #[serde(rename = "item.completed")]
    ItemCompleted {
        item: CodexItem,
    },
    /// 一轮对话完成
    #[serde(rename = "turn.completed")]
    TurnCompleted {
        usage: Option<CodexUsage>,
    },
}

/// codex 输出的项目（消息、工具调用、错误等）
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum CodexItem {
    /// 错误信息
    Error {
        id: String,
        message: String,
    },
    /// Agent 回复消息
    AgentMessage {
        id: String,
        text: String,
    },
    /// 工具调用
    ToolCall {
        id: String,
        name: String,
        arguments: Option<serde_json::Value>,
        status: Option<String>,
    },
    /// 审批请求
    ApprovalRequest {
        id: String,
        command: Option<String>,
        description: Option<String>,
    },
    /// 其他未识别的项目类型（兜底）
    #[serde(untagged)]
    Other(serde_json::Value),
}

/// token 使用统计
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodexUsage {
    pub input_tokens: Option<u64>,
    pub cached_input_tokens: Option<u64>,
    pub cache_write_input_tokens: Option<u64>,
    pub output_tokens: Option<u64>,
    pub reasoning_output_tokens: Option<u64>,
}

impl CodexJsonEvent {
    /// 从 JSON 字符串解析事件
    pub fn from_json(line: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(line)
    }
}
